src/thread_task.rs

```diff
@@ -813,869 +813,746 @@ unsafe fn delayed_fg_task(c: *const Rav1dContext, ttd: *mut TaskThreadData) {
-                                }
-                                (*ttd).cur = ((*ttd).cur).wrapping_add(1);
-                                current_block = 5601891728916014340;
-                            } else {
-                                if reset_task_cur(c, ttd, u32::MAX) != 0 {
-                                    continue 's_18;
-                                }
-                                if merge_pending(c) != 0 {
-                                    continue 's_18;
-                                } else {
-                                    current_block = 14728000373531839883;
-                                    break;
-                                }
+                        let frame_hdr = &***(*f).frame_hdr.as_ref().unwrap();
+                        let tile_row_base =
+                            frame_hdr.tiling.cols * (*f).frame_thread.next_tile_row[p as usize];
+                        if p != 0 {
+                            let p1_0 = (*f).frame_thread.entropy_progress.load(Ordering::SeqCst);
+                            if p1_0 < (*t).sby {
+                                // goto next;
+                                prev_t = t;
+                                t = (*t).next;
+                                (*f).task_thread.task_cur_prev = prev_t;
+                                continue;
                             }
+                            ::core::intrinsics::atomic_or_seqcst(
+                                &mut (*f).task_thread.error,
+                                (p1_0 == TILE_ERROR) as c_int,
+                            );
                         }
-                        _ => {
-                            if !prev_t.is_null() {
-                                (*prev_t).next = (*t).next;
-                            } else {
-                                (*f).task_thread.task_head = (*t).next;
-                            }
-                            if ((*t).next).is_null() {
-                                (*f).task_thread.task_tail = prev_t;
+                        let frame_hdr = &***(*f).frame_hdr.as_ref().unwrap();
+                        for tc_0 in 0..frame_hdr.tiling.cols {
+                            let ts: *mut Rav1dTileState = &mut *((*f).ts)
+                                .offset((tile_row_base + tc_0) as isize)
+                                as *mut Rav1dTileState;
+                            let p2 = ::core::intrinsics::atomic_load_seqcst(
+                                &mut *((*ts).progress).as_mut_ptr().offset(p as isize)
+                                    as *mut atomic_int,
+                            );
+                            if p2 < (*t).recon_progress {
+                                // goto next;
+                                prev_t = t;
+                                t = (*t).next;
+                                (*f).task_thread.task_cur_prev = prev_t;
+                                continue 'decoding;
```

Comment:
I think this should be a `continue` for the inner `while !t.is_null()` (yet to be labeled) loop, as the other `goto next`s are.

Reply:
Also, instead of inlining `goto next` everywhere, I think you can have a labeled block inside the `while !t.is_null()` loop, e.x. `'next`, use `break 'next` for the `goto next`s, and have the `next:` code following the `'next` block.  Or maybe name the `'next` block `'before_next`, and then have a `'_next'` block after for the `next:` code.

Reply:
Good catch. Surprised this didn't actually make a difference in testing. And yeah, I like the idea of a block, I'll add that.